### ISSUE TITLE
Allow mpOpenAPI to support multiple applications and modules

### ISSUE THREAD

**Issue by tevans78** [OpenLiberty/open-liberty#16628]
See: https://github.com/OpenLiberty/open-liberty/issues/15032

Currently, mpOpenAPI-2.0 only processes one web module in the first application to be deployed. This appears to be a limitation carried over from mpOpenAPI-1.1.

Code exists to merge OpenAPI documents from different applications in the `openapi-3.1` feature (which depends on `mpOpenAPI-1.0`) but this functionality is not available for `mpOpenAPI-2.0`.

As-is:
    one .war is deployed - an OpenAPI document is generated for that .war
    two .wars are deployed - an OpenAPI document is generated for one of the .wars
    an .ear containing two .wars is deployed - an OpenAPI document is generated for one of the .wars in the .ear

To-be:
   A single OpenAPI document is generated for all applications and modules
   Individual OpenAPI documents are still available for individual applications

<br/><br/><br/>  

##
#### Before proceeding to any items below (active development), this feature must be prioritized on the backlog, and have either the "Design Approved" or "No Design Approved" labels.  Follow the Feature and UFO Approval Process.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/8feimwmj8xzjibjept2s5hb7faxen2xl
- FTS: #17700
- Beta Blog Post (if applicable): #19103 
- Blog Post:  #19558
- Docs Issue: OpenLiberty/docs#4398

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - (Brian Hanczaryk - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by Emily-Jiang** [comment]
Recording is [here](https://ibm.ent.box.com/file/833668197047?s=nf9imo5396q6luvkgh96zlm7uc2nbpf2)

UFO review comments:

1. Mention merge conflict renaming rules for id and component
2. On FAT test slide, investigate the FAT suite for openapi-3.1 and execute these against mpOpenAPI-2.0 to work out whethere there is any difference between the two merging policies. 
3. Add the property that enables merging to the beta slide.

**Comment by Azquelt** [comment]
@Emily-Jiang Reviewing the openapi-3.1 tests, I couldn't see any FAT tests which would still apply, but there were a bunch of unit tests which could be adapted. This revealed the following differences:
* If the context root is `/`, mpOpenApi-2.0 creates paths which start `//`
  * Not sure if this happens in practise but it's easy enough to remove a trailing `/` from the context root
  * #19153
* openapi-3.1 corrects clashing names within the same document, mpOpenApi-2.0 does not
  * This cannot be done correctly as all the references within a document must be updated when a name changes. We can't do this if there are ambiguous references to start with.
* operation ids are renamed differently
  *  openapi-3.1 names them `operationid` `operationid_0` `operationid_1`
  * mpOpenApi-2.0 names them `operationid` `operationid1` `operationid2`
    * the mpOpenApi-2.0 scheme is consistent with the way both feature rename components
* If there are clashing paths between two applications, openapi-3.1 allows one to overwrite the other, mpOpenApi-2.0 reports an error
* mpOpenApi-2.0 does not intelligently handle a `server` with a relative URL
  * #19154

**Comment by Emily-Jiang** [comment]
@Azquelt please document the difference in the doc. It would be good to fix the issues you raised.

**Comment by Azquelt** [comment]
@Emily-Jiang Not sure how/where to document that. We don't have an existing page for differences between the old openapi features and mpOpenAPI. I strongly dislike having some changes between versions documented and others not.

**Comment by Emily-Jiang** [comment]
maybe just document the behaviour in the mpOpenAPI doc

**Comment by Azquelt** [comment]
@samwatibm This feature is complete and ready for Globalization review. Most new messages were delivered in #17697, however a new error condition was found at the end of development and a new message was added in #19218, missing the NL cutoff. Though I don't expect many people will hit this error case, I'm not sure if this means the feature has to be delayed until next release.

**Comment by Azquelt** [comment]
@steven1046 This feature is complete and ready for Accessibility review. There are no UI changes in this feature and I don't think there are any accessibility concerns.

**Comment by Azquelt** [comment]
@jhanders34 This feature is complete and ready for performance review. I sent a note to @jdmcclur earlier this week about it.

**Comment by Azquelt** [comment]
@hanczaryk This feature is complete and ready for SVT review.

**Comment by Azquelt** [comment]
@donbourne This feature is complete and ready for Serviceability review.

**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

    - Yes (slide 32).
    - We anticipate errors in configuration, for which we output appropriate log messages.
    - It's possible for documentation for different applications to clash while being merged, for which we output a log message explaining which documents clashed and where in the document the clash occurred.

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).

    a) What problem paths were tested and demonstrated?
      - Two applications are deployed which contain the same path, resulting in a clash and one application being omitted from the merged documentation
      - Two applications are deployed with JAX-RS resources but merging is not enabled so only one application is included in the documentation
      - An application is configured for inclusion in the merged documentation but no application with that name is deployed

    b) Who did you demo to?
      - @benjamin-confino (CDI & MicroProfile L3, not involved with the development of this feature)

    c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
      - Yes

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
    a) Who conducted SVT tests for this feature?
      - N/A

    b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
      - N/A

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
    - WASCDI

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
    - No new metrics or events

**Comment by Azquelt** [comment]
@chirp1 This feature is complete and ready for ID review. There is one docs issue open which I think has all the required information: OpenLiberty/docs#4398

**Comment by Azquelt** [comment]
@tevans78 I've added a demo to the agenda for the next EOI.

**Comment by chirp1** [comment]
Andrew provided the info that David needs in this issue: https://github.com/OpenLiberty/docs/issues/4398 . Approved.

**Comment by dave-waddling** [comment]
Hi,

I've granted the FAT Focal approval. The Feature Test Summary passes review and all testing is delivered and passing.

**Comment by jhanders34** [comment]
Granting perf focal approval.  While testing the feature though @jdmcclur ran into #19375 which we believe should be fixed before making this feature generally available.

**Comment by Azquelt** [comment]
Delivered in 22.0.0.1
